In [48]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [49]:
accd = pd.read_csv('datasets\\accident_data.csv')

In [50]:
accd

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [51]:
accd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [52]:
accd.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1>Clean the Dataset</h1>

In [53]:
accd['Latitude'] = accd['Latitude'].fillna(accd['Latitude'].mode()[0])
accd['Longitude'] = accd['Longitude'].fillna(accd['Longitude'].mode()[0])
accd['Road_Surface_Conditions'] = accd['Road_Surface_Conditions'].fillna(accd['Road_Surface_Conditions'].mode()[0])
accd['Road_Type'] = accd['Road_Type'].fillna(accd['Road_Type'].mode()[0])
accd['Urban_or_Rural_Area'] = accd['Urban_or_Rural_Area'].fillna(accd['Urban_or_Rural_Area'].mode()[0])
accd['Weather_Conditions'] = accd['Weather_Conditions'].fillna(accd['Weather_Conditions'].mode()[0])

In [54]:
accd.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [55]:
accd['Accident_Severity'] = accd['Accident_Severity'].astype('category')
accd['Accident Date'] = accd['Accident Date'].astype('category')
accd['Latitude'] = accd['Latitude'].astype('category')
accd['Light_Conditions'] = accd['Light_Conditions'].astype('category')
accd['Longitude'] = accd['Longitude'].astype('category')
accd['Road_Surface_Conditions'] = accd['Road_Surface_Conditions'].astype('category')
accd['Road_Type'] = accd['Road_Type'].astype('category')
accd['Urban_or_Rural_Area'] = accd['Urban_or_Rural_Area'].astype('category')
accd['Weather_Conditions'] = accd['Weather_Conditions'].astype('category')
accd['Vehicle_Type'] = accd['Vehicle_Type'].astype('category')

In [56]:
accd.dtypes

Index                        object
Accident_Severity          category
Accident Date              category
Latitude                   category
Light_Conditions           category
District Area                object
Longitude                  category
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object

In [57]:
accd['Accident Date'] = pd.to_datetime(accd['Accident Date'],dayfirst = True, errors = 'coerce')

In [58]:
accd['Year']=accd['Accident Date'].dt.year
accd['Month']=accd['Accident Date'].dt.month
accd['Day']=accd['Accident Date'].dt.day
accd['DayOfWeek']=accd['Accident Date'].dt.dayofweek #Monday=0, Sunday=

<h1>Questions and Insights</h1>

<h1>1.Which district has the highest average per accident?</h1>

In [59]:
accd.groupby('District Area')['Number_of_Casualties'].mean().sort_values(ascending=False).head()

District Area
Blaeu Gwent          1.693694
West Somerset        1.613990
Derwentside          1.606186
Chester-le-Street    1.603306
Durham               1.580431
Name: Number_of_Casualties, dtype: float64

<h2>Insight 1: The highest average accident district area is Blaeu Gwent with 1.693694 </h2>

<h1>2. What is the most common accident severity?</h1>

In [60]:
accd['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>Insight 2: The most common accident severity is Slight with 563801</h2>

<h1>3. Which light condition has the highest number of accidents? </h1>

In [61]:
accd['Light_Conditions'].value_counts().idxmax()

'Daylight'

<h2>Insight 3: Daylight  has the highest number of accidents in light condition</h2>

<h1>4.Which road surface condition is linked to the highest number of accidents?</h1>

In [62]:
accd['Road_Surface_Conditions'].value_counts().idxmax()

'Dry'

<h2>Insight 4: Dry got the highest number of accident on road surface condition</h2>

<h1>5. Which year recorded the highest number of accidents?</h1>

In [63]:
accident = accd['Year'].value_counts().idxmax()

In [64]:
accident

np.float64(2019.0)

<h2>Insight 5:Based on the results the year recorded highest number of accidents is 2019 </h2>

<h1>6. Is there a correlation between the number of vehicles involved and the number of casualties?</h1>

In [65]:
accd[['Number_of_Vehicles', 'Number_of_Casualties']].corr().iloc[0,1]

np.float64(0.22888886126926722)

<h1>Insight 6: There is no correlation between number vehicles involed and the number of casualties</h1>

<h1>7.Which month across all years has the highest accident frequency?</h1>

In [66]:
 accd['Month'].value_counts().idxmax()

np.float64(11.0)

<h2> Insight 7: 11 or November got the highest accident frequency across all years</h2>

<h1>8. What kind of weather conditions have the highest record of accident record?</h1>

In [67]:
weather = accd['Weather_Conditions'].value_counts()

In [68]:
weather

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>Insight 8: Fine no high winds got the highest record of accident </h2>

<h1>9.Which date has the most recorded of accident?</h1>

In [69]:
accidents = accd['Accident Date'].value_counts()

In [70]:
accidents

Accident Date
2021-02-11    685
2021-10-06    680
2019-12-06    678
2019-02-01    676
2021-12-04    667
             ... 
2022-01-09    236
2022-12-02    236
2022-04-05    231
2022-02-07    213
2022-01-10    123
Name: count, Length: 576, dtype: int64

<h2>Insight 9: 2021-02-11 has the most recorded accident date with 685 accidents</h2>

<h1>10.Is there a correlation between the Latitude and the number of casualties that is recorded in the dataset?</h1>

In [71]:
corr_lat_cas = accd['Latitude'].corr(accd['Number_of_Casualties'])

In [72]:
corr_lat_cas

np.float64(0.03220129464996815)

<h2>Insight 10: According to the result of the analysis there is no correlation between Latitude and Number of casualties in the dataset</h2>

<h1>11. Which light condition has the highest proportion of fatal accidents?</h1>

In [73]:
fatal_ratio = accd[accd['Accident_Severity']=="Fatal"].groupby("Light_Conditions").size() / accd.groupby("Light_Conditions").size()

In [74]:
fatal_ratio 

Light_Conditions
Darkness - lighting unknown    0.010487
Darkness - lights lit          0.014381
Darkness - lights unlit        0.017696
Darkness - no lighting         0.043059
Daylight                       0.010469
dtype: float64

<h2>Insight 11: Based on the result the highest proportion of fatal accidents on light conditions is Darkness - no lighting</h2>

<h1>12.Which district area recorded the fewest accidents?</h1>

In [82]:
accd['District Area'].value_counts().idxmin()


'Clackmannanshire'

<h2>Insight 12: Based on the result Clackmannashire is the fewest accident recorded in district area</h2>

<h1>13. Do accidents involving motorcycles result in more casualties on average than cars?</h1>

In [80]:
avg_casualties = accd.groupby("Vehicle_Type")["Number_of_Casualties"].mean()

In [81]:
avg_casualties

Vehicle_Type
Agricultural vehicle                     1.342065
Bus or coach (17 or more pass seats)     1.349216
Car                                      1.358841
Data missing or out of range             1.500000
Goods 7.5 tonnes mgw and over            1.351881
Goods over 3.5t. and under 7.5t          1.362861
Minibus (8 - 16 passenger seats)         1.345648
Motorcycle 125cc and under               1.332635
Motorcycle 50cc and under                1.337235
Motorcycle over 125cc and up to 500cc    1.343521
Motorcycle over 500cc                    1.359434
Other vehicle                            1.340075
Pedal cycle                              1.370558
Ridden horse                             1.250000
Taxi/Private hire car                    1.368663
Van / Goods 3.5 tonnes mgw or under      1.354537
Name: Number_of_Casualties, dtype: float64

<h2>Insight 13: Overall, motorcycle accidents do not consistently result in more casualties than car accidents.</h2>

<h1>14.Which road surface condition is associated with the lowest average number of casualties? </h1>

In [83]:
accd.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean().idxmin()

'Dry'

<h2>Insight 14: According to the data result the lowest average number of casualties on road surface condition is Dry </h2>

<h1>15. Which year shows the largest year-over-year decrease in accidents? </h1>

In [86]:
yearly_counts = accd['Year'].value_counts().sort_index()

In [87]:
yearly_counts

Year
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

<h2>Insight 15:Based on the result the largest year-over-year decrease in accidents is 2019</h2>

<h1>16.Do accidents involving buses/coaches have more casualties on average than cars?</h1>

In [90]:
avg_casualties = accd.groupby("Vehicle_Type")["Number_of_Casualties"].mean()

In [91]:
avg_casualties

Vehicle_Type
Agricultural vehicle                     1.342065
Bus or coach (17 or more pass seats)     1.349216
Car                                      1.358841
Data missing or out of range             1.500000
Goods 7.5 tonnes mgw and over            1.351881
Goods over 3.5t. and under 7.5t          1.362861
Minibus (8 - 16 passenger seats)         1.345648
Motorcycle 125cc and under               1.332635
Motorcycle 50cc and under                1.337235
Motorcycle over 125cc and up to 500cc    1.343521
Motorcycle over 500cc                    1.359434
Other vehicle                            1.340075
Pedal cycle                              1.370558
Ridden horse                             1.250000
Taxi/Private hire car                    1.368663
Van / Goods 3.5 tonnes mgw or under      1.354537
Name: Number_of_Casualties, dtype: float64

<h2>Insight 16: The average casualties per bus accident (1.35) is actually slightly lower than for cars (1.36).</h2>

<h1>17.Is there a correlation between weather conditions and accident severity?</h1>

In [100]:
severity_by_weather = accd.groupby('Weather_Conditions')['Accident_Severity'].value_counts(normalize=True).unstack()

In [101]:
severity_by_weather 

Accident_Severity         Fatal   Serious    Slight
Weather_Conditions                                 
Fine + high winds      0.020458  0.145546  0.833996
Fine no high winds     0.013471  0.136978  0.849551
Fog or mist            0.023243  0.136905  0.839853
Other                  0.009621  0.105015  0.885364
Raining + high winds   0.015081  0.131149  0.853770
Raining no high winds  0.010640  0.118801  0.870558
Snowing + high winds   0.003390  0.123164  0.873446
Snowing no high winds  0.005771  0.090574  0.903655

<h2>Insight 17: Base on the result there is a correlation between weather condition and accident severity</h2>

<h1>18.Does the average number of vehicles involved per accident differ between urban and rural areas?</h1>

In [111]:
avg_vehicles_area = accd.groupby("Urban_or_Rural_Area")["Number_of_Vehicles"].mean()

In [112]:
avg_vehicles_area

Urban_or_Rural_Area
Rural          1.866312
Unallocated    1.909091
Urban          1.811385
Name: Number_of_Vehicles, dtype: float64

<h2>Insight 18: Rural areas generally see larger multi-vehicle accidents than urban areas.</h2>

<h1>19.Which area type shows the most variation (spread) in number of vehicles?</h1>

In [120]:
accd.groupby("Urban_or_Rural_Area")["Number_of_Vehicles"].std()

Urban_or_Rural_Area
Rural          0.809018
Unallocated    0.539360
Urban          0.655390
Name: Number_of_Vehicles, dtype: float64

<h2>Insight 19:Rural has the most variation in number of vehicles </h2>

<h1>20.Which vehicle type is most frequently involved in accidents in rural areas?</h1>

In [129]:
rural_vehicle_counts = accd[accd["Urban_or_Rural_Area"] == "Rural"]["Vehicle_Type"].value_counts().head(5)

In [130]:
rural_vehicle_counts

Vehicle_Type
Car                                     181922
Van / Goods 3.5 tonnes mgw or under      12156
Bus or coach (17 or more pass seats)      9025
Motorcycle over 500cc                     8957
Goods 7.5 tonnes mgw and over             6156
Name: count, dtype: int64

<h2>Insight 20:Car got the most frequently involved in accident in rural areas </h2>

<h1>21.Do rainy weather conditions result in more casualties compared to fine weather?</h1>

In [126]:
rain_vs_fine = accd.groupby("Weather_Conditions")["Number_of_Casualties"].mean()

In [127]:
rain_vs_fine

Weather_Conditions
Fine + high winds        1.386018
Fine no high winds       1.347397
Fog or mist              1.452948
Other                    1.354869
Raining + high winds     1.416641
Raining no high winds    1.408214
Snowing + high winds     1.418079
Snowing no high winds    1.341776
Name: Number_of_Casualties, dtype: float64

<h2>Insight 21: Based on the result rain weather got a more casualties compared to fine</h2>

<h1>22.Do rural districts have higher average casualties than urban districts?</h1>

In [133]:
casualties_area = accd.groupby("Urban_or_Rural_Area")["Number_of_Casualties"].mean()

In [134]:
casualties_area

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h2>Insight 22: Rural district accidents are more severe than Urban district</h2>

<h1>23.What is the most common number of vehicles involved in accidents?</h1>

In [135]:
accd['Number_of_Vehicles'].value_counts().idxmax()

np.int64(2)

<h2>Insight 23:According to the result 2 is the number of vehicle commonly involved in accident.</h2>

<h1>24.How has the number of accidents on each Road_Type changed over the years?</h1>

In [139]:
accidents_by_year_road = accd.groupby(['Year','Road_Type']).size().unstack(fill_value=0)

In [140]:
accidents_by_year_road

Road_Type  Dual carriageway  One way street  Roundabout  Single carriageway  \
Year                                                                          
2019.0                11022            1550        4603               53917   
2020.0                10527            1431        4688               52776   
2021.0                 9764            1321        4408               49981   
2022.0                 8319            1166        3831               42881   

Road_Type  Slip road  
Year                  
2019.0           775  
2020.0           741  
2021.0           698  
2022.0           608

<h2>Insight 24:Based on the result as the years past the number of accident decreases</h2>

<h1>25.Do accidents with more vehicles tend to have higher casualty counts?</h1>

In [137]:
accd.groupby('Number_of_Vehicles')['Number_of_Casualties'].mean()

Number_of_Vehicles
1      1.170932
2      1.374880
3      1.711169
4      1.995575
5      2.315341
6      2.612137
7      3.064189
8      3.401361
9      3.350877
10     3.629630
11     4.000000
12     2.285714
13     7.833333
14     5.444444
15     5.000000
16     6.000000
19    13.000000
28    16.000000
32     5.000000
Name: Number_of_Casualties, dtype: float64

<h2>Insight 25:Yes, as the number of vehicles increases, average casualties rise.</h2>